CONTRIBUTING.md

```diff
@@ -255,6 +255,49 @@ pub use {
 }
 ```
 
+#### Return `Self`
+
+Use `Self` as the return type instead of naming the type.
+
+```rust
+/// A counter that is always smaller than 100.
+pub struct Counter(u32);
+
+impl Counter {
+    /// Constructs a new `Counter`.
+    pub fn new() -> Self { Self(0) }
+
+    /// Returns a counter if it is possible to create one from x.
+    pub fn maybe(x: u32) -> Option<Self> {
+        match x {
+            x if x >= 100 => None,
+            c => Some(Counter(c)),
+        }
+    }
+}
+
+impl TryFrom<u32> for Counter {
+    type Error = TooBigError;
+
+    fn try_from(x: u32) -> Result<Self, Self::Error> {
+        if x >= 100 {
+            return Err(TooBigError);
+        }
+        Ok(Counter(x))
+    }
+}
+```
+
+When constructing the return value use `Self` or the actual type, as you prefer.
+
+```rust
+impl From<some::Error> for Error {
+    fn from(e: some::Error) -> Self { Self::Some(e) }
```

Comment:
In aab29fcfc10306754b1c9f3458b14089a9bbc08b:

Is this `Self::Some` syntax actually allowed? A long time ago it was illegal to use enum variants off of newtypes.

Reply:
Yeah its allowed, I've been using it for a while now in error cleanups. see `bitcoin/src/address/error.rs` for an example 

```rust
impl From<base58::Error> for ParseError {
    fn from(e: base58::Error) -> Self { Self::Base58(e) }
}
```

Reply:
I suspect this is why the lint has MSRV 1.29